enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/block_permissions.clj

```diff
@@ -9,6 +9,11 @@
   [database-or-id]
   (contains? @api/*current-user-permissions-set* (perms/database-block-perms-path database-or-id)))
 
+(defn- current-user-has-native-query-data-perm-despite-having-block-perm-for-database?
+  [database-or-id]
+  (and (current-user-has-block-permissions-for-database? database-or-id)
+       (contains? @api/*current-user-permissions-set* (perms/all-schemas-path database-or-id))))
```

Comment:
I think you meant `perms/adhoc-native-query-path` here

Reply:
that would be the read and write perm, which they explicitly said was the one they didn't want

Reply:
I don't know what you guys are talking about when you're saying `perms/adhoc-native-query-path` means "write" permissions. `adhoc-native-query-path` _is_ the permission to run read-only* ad-hoc native queries and that's it. `all-schemas-path` is **specifically** for MBQL queries only. We should not be giving people permissions to run ad-hoc native queries based on an MBQL-specific permission no matter what, and anyone who is telling you to do otherwise doesn't understand how this stuff works.

`*` = we try to execute all queries in read-only mode but it's a best-effort thing rather than a guarantee and it depends on the  DB

At any rate you don't want to be doing `contains?` here to look for a specific permission because our perms stuff is set up with a prefix system.

If I have perms for `/` (superuser perms) I have perms for anything starting with `/`. If I have perms for `/database/1/` I have perms for everything related to Database 1...  There are functions in `metabase.models.permissions` for checking whether a given permissions set should be considered to have permissions for a given path or not. So you should use these

Reply:
I think part of the confusion here is that _ad-hoc native query running permissions_ is labeled as _native query editing_ in the UI these days. That label is completely misleading. That's not what setting that permission does at all. Whether or not you can **edit** a Saved Question (aka Card) depends on whether or not you have write ("curate") permissions for the Collection it lives in. Whether you can run an ad-hoc native query depends on _ad-hoc native query running permissions_